components/UserDashboard.tsx

```diff
@@ -38,41 +53,74 @@ const UserDashboardItem: React.FunctionComponent<User> = ({
 
 const UserDashboard: React.FunctionComponent = () => {
   const [users, setUsers] = useState<User[]>();
-
-  const getUsers = async (): Promise<void> => {
-    try {
-      const response = await fetch("http://localhost:3000/api/admin/users", {
-        method: "GET",
-        headers: { "content-type": "application/json" },
-        redirect: "follow",
-      });
-      const data = await response.json();
-      setUsers(data.users);
-    } catch (err) {
-      throw new Error(err.message);
-    }
-  };
+  const [index, setIndex] = useState(0);
+  const [uiPage, setUIPage] = useState(0);
+  const [numUIPages, setNumUIPages] = useState(0);
+  const [pageCache, setPageCache] = useState<Record<number, User[]>>({});
 
   useEffect(() => {
-    getUsers();
-  }, []);
+    const getUsers = async (pageNumber: number): Promise<void> => {
+      try {
+        if (pageNumber in pageCache) {
+          // Page already in cache; no need to make a request!
+          setUsers(pageCache[pageNumber]);
+        } else {
+          const response = await fetch(
+            `http://localhost:3000/api/admin/users?pageNumber=${pageNumber}`,
+            {
+              method: "GET",
+              headers: { "content-type": "application/json" },
+              redirect: "follow",
+            }
+          );
+          const data = await response.json();
+          setUsers(data.users);
+          setNumUIPages(Math.ceil(data.total / UI_PAGE_SIZE));
+          setPageCache({
+            ...pageCache,
+            [pageNumber]: data.users,
+          });
+        }
+      } catch (err) {
+        throw new Error(err.message);
+      }
```

Comment:
Rethrowing here is the same as not wrapping this in a try/catch! Since we don't have an error state designed yet, how about we drop the try/catch and include a TODO about error handling for the future?